File: IncreasingSubsequences2.cpp
#include<iostream>
#include<vector>
#include<unordered_set>

using namespace std;

class Solution {
public:
    vector<vector<int> > findSubsequences(vector<int>& nums) {
        vector<vector<int> > result;
        vector<int> sequence;
        findSubSequence(result, nums, 0, sequence);
        return result;
    }
    
    void findSubSequence(vector<vector<int> >& result, vector<int>& nums, int pos, vector<int>& sequence) {
    	if(sequence.size() >= 2) result.push_back(sequence);
    	unordered_set<int> s;
    	for(int i = pos; i < nums.size(); i++) {
    		if(s.find(nums[i]) != s.end()) continue;
    		s.insert(nums[i]);
    		if(sequence.empty() || nums[i] >= sequence.back()) {
    			sequence.push_back(nums[i]);
    			findSubSequence(result, nums, i + 1, sequence);
    			sequence.pop_back();
			}
		}
	}
};

int main() {
	int n;
	cin>>n;
	
	vector<int> nums;
	for(int i = 0;i < n; i++) {
		int num;
		cin>>num;
		nums.push_back(num);
	}
	
	Solution *solution = new Solution();
	vector<vector<int> > result = solution->findSubsequences(nums);
	
	for(int i = 0; i < result.size(); i++) {
		for(int j = 0; j < result[i].size(); j++) cout<<result[i][j]<<"  ";
		cout<<endl;
	}
}

File: IncreasingSubsequences2.py
class Solution:
    def findSubsequences(self, nums: list) -> list:
        result, cur = [], [[]]
        m = {}
        for i in range(len(nums)):
            start = m.get(nums[i], 0)
            n = len(cur)
            m[nums[i]] = n
            for j in range(start, n, 1):
                if not cur[j] or cur[j][-1] <= nums[i]:
                    cur.append(cur[j].copy())
                    cur[-1].append(nums[i])
                    if len(cur[-1]) >= 2:
                        result.append(cur[-1])
        return result


if __name__ == '__main__':
    nums = [int(num) for num in input().split()]
    solution = Solution()
    print(solution.findSubsequences(nums))
